### ISSUE TITLE
Unifi UXG Pro does not work

### ISSUE THREAD

**Issue by mbnn** [nextdns/nextdns#476]
When trying to install it on the Unifi UXG Pro router i get this message (it's based on the same kernel as UDM (Pro)):

```
# sh -c 'sh -c "$(curl -sL https://nextdns.io/install)"'
INFO: OS: ubios
INFO: GOARCH: arm64
INFO: GOOS: linux
INFO: NEXTDNS_BIN: /data/nextdns
INFO: LATEST_RELEASE: 1.32.0
i) Install NextDNS
e) Exit
Choice (default=i): i
INFO: Installing NextDNS...
Error: invalid arguments /tmp/nextdns.list, unifi-os:/etc/apt/sources.list.d/nextdns.list you must use just one container
Error: no container with name or ID unifi-os found: no such container
Error: no container with name or ID unifi-os found: no such container
Error: no container with name or ID unifi-os found: no such container
```

Is there anything specific I can test/do to fix this? :)
Running UXG Pro firmware 1.9.0-4.

**Comment by rs** [comment]
UXG Pro is not supported yet.

**Comment by mbnn** [comment]
Ok, thanks.

**Comment by jrusbatch** [comment]
I also came here to report this. Is support for the UXG Pro planned for an upcoming release? I really enjoyed the peace of mind that running NextDNS on my old USG gave me.

**Comment by mbnn** [comment]
@rs anything we can do to help? :)

**Comment by rs** [comment]
What do you get for `podman ps` on the unit?

**Comment by mbnn** [comment]
Output:

```
# podman ps
CONTAINER ID  IMAGE                       COMMAND               CREATED      STATUS          PORTS  NAMES
8c2e26075002  localhost/uxg-setup:latest  nodejs /usr/share...  4 weeks ago  Up 4 weeks ago         uxg-setup
```

**Comment by mbnn** [comment]
@rs Any idea? 😀

**Comment by mbnn** [comment]
Guess not :P

**Comment by sdjernes** [comment]
Is anyone working on this? Or, is this dead because Ubiquity is not shipping product yet.  I have been waiting over six months to try to buy one, as my USG-4-Pro is massively underpowered to do IDS/IPS on a 1Gig Fiber connection.

**Comment by mbnn** [comment]
I went over to the UDM Pro SE, which works 😀

**Comment by sdjernes** [comment]
The UDM Pro / Pro SE is really overkill for just routing for me. I already have 2 48-port switches, the UNVR-Pro and the controller running in a container on my QNAP.

**Comment by unsaltedorg** [comment]
Fellow UXG owner here, experiencing the same error. The UXG Pro is now GA so there may be more people running into this.

I added a $75 bounty for this issue [on BountySource](https://app.bountysource.com/issues/98082586-unifi-uxg-pro-does-not-work) (!). 

It would be amazing if someone could reopen this and take a look.

**Comment by curiositycasualty** [comment]
Poking around with some of the install script's `FORCE_*` ENVs to no avail. Even if the installation seems successful, running the `nextdns` binary (for any of the subcommands besides `version`) results in: `system not supported`. On the plus side, that means the architecture of the binary is correct and that it executes. I think it's tripping over some of the checks for the `ubios` platform like: https://github.com/nextdns/nextdns/blob/90dca2ec199308d69578dc5cdfd99046a805ba72/host/service/ubios/service.go#L28-L30

I've included a ton of info from my UXG Pro. I've also got 2x uSgs at my disposal. Let me know if there anything else I can help with.

It's unclear to me what the analog for some of these checks would be for the UXG Pro's UbiOS as it seems novel compared to the USG.

https://github.com/nextdns/nextdns/blob/90dca2ec199308d69578dc5cdfd99046a805ba72/host/service/ubios/service.go#L45

`# cat /proc/1/cgroup`
```
2:devices:/
1:name=systemd:/
0::/
```

`# podman ps`
```
CONTAINER ID  IMAGE                       COMMAND               CREATED       STATUS         PORTS  NAMES
2f512bd9b148  localhost/uxg-setup:latest  nodejs /usr/share...  3 months ago  Up 2 days ago         uxg-setup
```

`# podman image ls`
```
REPOSITORY            TAG       IMAGE ID       CREATED        SIZE     R/O
localhost/uxg-setup   latest    906259128924   3 months ago   316 MB   true
localhost/uxg-setup   default   906259128924   3 months ago   316 MB   true
```

`# podman info`
```
host:
  BuildahVersion: 1.11.2
  CgroupVersion: v2
  Conmon:
    package: Unknown
    path: /usr/libexec/podman/conmon
    version: 'conmon version 2.0.1, commit: 890392262a258b98e49f94827e150ab55b758914'
  Distribution:
    distribution: ubios
    version: v1.11.0.3923-8903922
  MemFree: 859222016
  MemTotal: 2096848896
  OCIRuntime:
    package: Unknown
    path: /usr/bin/runc
    version: |-
      runc version commit: v1.0.0-rc6
      spec: 1.0.1-dev
  SwapFree: 3669483520
  SwapTotal: 3669483520
  arch: arm64
  cpus: 4
  eventlogger: file
  hostname: UXGPro
  kernel: 4.19.152-al-linux-v10.2.0-v1.11.0.3923-8903922
  os: linux
  rootless: false
  uptime: 59h 8m 19.72s (Approximately 2.46 days)
registries:
  blocked: null
  insecure: null
  search:
  - docker.io
  - quay.io
  - registry.fedoraproject.org
store:
  ConfigFile: /etc/containers/storage.conf
  ContainerStore:
    number: 1
  GraphDriverName: overlay
  GraphOptions:
    .imagestore: /var/lib/containers/storage
    .mountopt: nodev
    .skip_mount_home: "false"
  GraphRoot: /mnt/data/podman/storage
  GraphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  ImageStore:
    number: 1
  RunRoot: /var/run/containers/storage
  VolumePath: /mnt/data/podman/storage/volumes
```

`# uname -a`
```
Linux UXGPro 4.19.152-al-linux-v10.2.0-v1.11.0.3923-8903922 #1 SMP Mon Dec 20 16:52:42 UTC 2021 aarch64 GNU/Linux
```

`# cat /etc/os-release`
```
NAME=UbiOS
VERSION=UXG.alpinev2.v1.11.0.8903922.211220.1547
ID=ubios
VERSION_ID=v1.11.0.3923-8903922
PRETTY_NAME="UbiOS 1.11.0.3923"
BUILD_ID=211220.1547
VARIANT=UXG.alpinev2
```

<details>
<summary> "sh -x -c 'DEBUG=1 sh -x ./install-nextdns.sh'" </summary>

```
+ DEBUG=1 sh -x ./install-nextdns.sh
+ umask 0022
+ main
+ detect_os
+ '['  ]
+ uname -s
+ uname -o
+ grep -q -e ^EdgeRouter -e ^UniFiSecurityGateway /etc/version
+ uname -u
+ grep -q ^synology
+ . /etc/os-release
+ NAME=UbiOS
+ VERSION=UXG.alpinev2.v1.11.0.8903922.211220.1547
+ ID=ubios
+ VERSION_ID=v1.11.0.3923-8903922
+ PRETTY_NAME='UbiOS 1.11.0.3923'
+ BUILD_ID=211220.1547
+ VARIANT=UXG.alpinev2
+ echo ubios
+ dist=ubios
+ command -v podman
+ '[' -z /usr/bin/podman ]
+ echo ubios
+ return 0
+ OS=ubios
+ detect_goarch
+ '['  ]
+ uname -m
+ uname -o
+ echo arm64
+ GOARCH=arm64
+ detect_goos
+ '['  ]
+ uname -s
+ echo linux
+ GOOS=linux
+ bin_location
+ echo /data/nextdns
+ NEXTDNS_BIN=/data/nextdns
+ get_release
+ '['  ]
+ command -v curl
+ break
+ cmd='curl -A curl -s'
+ grep '"tag_name":'
+ curl+ sed -e s/^v//
 -A+  curlesed -s https://api.github.com/repos/nextdns/nextdns/releases/latest 's/.*"([^"]+)".*/\1/'

+ echo+ sed -E

+ sed -E 's/.*"([^"]+)".*/\1/'
+ v=1.37.11
+ '[' -z 1.37.11 ]
+ echo 1.37.11
+ INSTALL_RELEASE=1.37.11
+ export 'NEXTDNS_INSTALLER=1'
+ log_info 'OS: ubios'
+ printf 'INFO: %s\n' 'OS: ubios'
INFO: OS: ubios
+ log_info 'GOARCH: arm64'
+ printf 'INFO: %s\n' 'GOARCH: arm64'
INFO: GOARCH: arm64
+ log_info 'GOOS: linux'
+ printf 'INFO: %s\n' 'GOOS: linux'
INFO: GOOS: linux
+ log_info 'NEXTDNS_BIN: /data/nextdns'
+ printf 'INFO: %s\n' 'NEXTDNS_BIN: /data/nextdns'
INFO: NEXTDNS_BIN: /data/nextdns
+ log_info 'INSTALL_RELEASE: 1.37.11'
+ printf 'INFO: %s\n' 'INSTALL_RELEASE: 1.37.11'
INFO: INSTALL_RELEASE: 1.37.11
+ '[' -z ubios ]
+ '[' -z arm64 ]
+ '[' -z linux ]
+ '[' -z /data/nextdns ]
+ '[' -z 1.37.11 ]
+ true
+ get_current_release
+ '[' -x /data/nextdns ]
+ CURRENT_RELEASE=
+ log_debug 'Start install loop with CURRENT_RELEASE='
+ '[' 1 '=' 1 ]
+ printf '\033[30;1mDEBUG: %s\033[0m\n' 'Start install loop with CURRENT_RELEASE='
DEBUG: Start install loop with CURRENT_RELEASE=
+ '['  ]
+ log_debug 'NextDNS is not installed'
+ '[' 1 '=' 1 ]
+ printf '\033[30;1mDEBUG: %s\033[0m\n' 'NextDNS is not installed'
DEBUG: NextDNS is not installed
+ menu i 'Install NextDNS' install e Exit exit
+ true
+ n=0
+ default=
+ key=i
+ '[' -z  ]
+ default=i
+ n=1
+ echo 'i) Install NextDNS'
i) Install NextDNS
+ n=2
+ n=3
+ key=e
+ '[' -z i ]
+ n=4
+ echo 'e) Exit'
e) Exit
+ n=5
+ n=6
+ print 'Choice (default=%s): ' i
+ format='Choice (default=%s): '
+ '[' 2 -gt 0 ]
+ shift
+ printf 'Choice (default=%s): ' i
Choice (default=i): + read -r choice
i
+ '[' -z i ]
+ n=0
+ key=i
+ n=1
+ n=2
+ '[' i '=' i ]
+ install
+ get_current_release
+ '[' -x /data/nextdns ]
+ '['  ]
+ install_type
+ '['  ]
+ echo ubios
+ type=ubios
+ log_info 'Installing NextDNS...'
+ printf 'INFO: %s\n' 'Installing NextDNS...'
INFO: Installing NextDNS...
+ log_debug 'Using ubios install type'
+ '[' 1 '=' 1 ]
+ printf '\033[30;1mDEBUG: %s\033[0m\n' 'Using ubios install type'
DEBUG: Using ubios install type
+ install_ubios
+ ubios_install_source
+ echo 'deb [signed-by=/usr/share/keyrings/nextdns.gpg] https://repo.nextdns.io/deb stable main'
+ podman exec unifi-os mv /data/nextdns.list /etc/apt/sources.list.d/nextdns.list
Error: no container with name or ID unifi-os found: no such container
+ rm -f /tmp/nextdns.list
+ podman exec unifi-os apt-get install -y gnupg1 curl
Error: no container with name or ID unifi-os found: no such container
+ podman exec unifi-os curl -sfL https://repo.nextdns.io/nextdns.gpg -o /usr/share/keyrings/nextdns.gpg
Error: no container with name or ID unifi-os found: no such container
+ podman exec unifi-os apt-get update -o 'Dir::Etc::sourcelist=sources.list.d/nextdns.list' -o 'Dir::Etc::sourceparts=-' -o 'APT::Get::List-Cleanup=0'
Error: no container with name or ID unifi-os found: no such container
+ podman exec unifi-os apt-get install -y nextdns
Error: no container with name or ID unifi-os found: no such container
+ break 2
+ true
+ get_current_release
+ '[' -x /data/nextdns ]
+ CURRENT_RELEASE=
+ log_debug 'Start install loop with CURRENT_RELEASE='
+ '[' 1 '=' 1 ]
+ printf '\033[30;1mDEBUG: %s\033[0m\n' 'Start install loop with CURRENT_RELEASE='
DEBUG: Start install loop with CURRENT_RELEASE=
+ '['  ]
+ log_debug 'NextDNS is not installed'
+ '[' 1 '=' 1 ]
+ printf '\033[30;1mDEBUG: %s\033[0m\n' 'NextDNS is not installed'
DEBUG: NextDNS is not installed
+ menu i 'Install NextDNS' install e Exit exit
+ true
+ n=0
+ default=
+ key=i
+ '[' -z  ]
+ default=i
+ n=1
+ echo 'i) Install NextDNS'
i) Install NextDNS
+ n=2
+ n=3
+ key=e
+ '[' -z i ]
+ n=4
+ echo 'e) Exit'
e) Exit
+ n=5
+ n=6
+ print 'Choice (default=%s): ' i
+ format='Choice (default=%s): '
+ '[' 2 -gt 0 ]
+ shift
+ printf 'Choice (default=%s): ' i
Choice (default=i): + read -r choice
e
+ '[' -z e ]
+ n=0
+ key=i
+ n=1
+ n=2
+ '[' i '=' e ]
+ n=3
+ key=e
+ n=4
+ n=5
+ '[' e '=' e ]
+ exit
```

</details>

**Comment by unsaltedorg** [comment]
@curiositycasualty Thank you so much for your investigations on this issue! 

Fingers crossed this will lead us to a fix.

**Comment by unsaltedorg** [comment]
@rs See above for super detailed debug info from @curiositycasualty  😄

**Comment by mbnn** [comment]
Nice job on all the investigations 😀  I hope they will pick it up.

**Comment by unsaltedorg** [comment]
@rs could this issue be re-opened please?

**Comment by unsaltedorg** [comment]
> @rs could this issue be re-opened please?

Thank you! 🙂

**Comment by unsaltedorg** [comment]
There is now a total of $225 pledged to solving this issue [on BountySource](https://app.bountysource.com/issues/98082586-unifi-uxg-pro-does-not-work).

**Comment by unsaltedorg** [comment]
Ryan Posner made the following progress (posted on a [NextDNS Community support thread](https://help.nextdns.io/t/35hbd1q/nextdns-install-fails-on-unifi-uxg-pro) yesterday):

> I got a little further with the install where I replaced “UniFi-us” with “uxg-setup” in the install file. I replaced the install url with my Dropbox link below when trying to install in ssh. Looks like there are read/write permissions blocking the creation of the directory and file needed to complete the install  

> https://www.dropbox.com/s/xhc05iu0of9hkki/install.txt?dl=0

**Comment by curiositycasualty** [comment]
> > I got a little further with the install where I replaced “UniFi-us” with “uxg-setup” in the install file.

Had this thought too, but I wasn't confident that the `uxg-setup` container was an equivalent of `UniFi-us` container :/
Glaringly, `uxg-setup` is mounted at read-only, which I think differentiates it from the previous gateway generations.

**Comment by unsaltedorg** [comment]
@rs What can I do to move this forward?

**Comment by mbnn** [comment]
Has anyone found a workaround for the UXG and NextDNS? :) Would love to have the CLI on there.

**Comment by mbnn** [comment]
> There is now a total of $225 pledged to solving this issue [on BountySource](https://app.bountysource.com/issues/98082586-unifi-uxg-pro-does-not-work).

Right now thats even more 😀 375, but no updates so far.
@rs Maybe you can help?

**Comment by mbnn** [comment]
No one? 😀

**Comment by mbnn** [comment]
@dotgrid , @rs any update by any chance, please? 😀

**Comment by mbnn** [comment]
I have no coding skills but Im willing to donate another 100$ it this can get fixed for the UXG 😀 Thanks.

**Comment by curiositycasualty** [comment]
@mbnn I wouldn't even classify this as a coding challenge at this point. I'd love to be wrong, but I think UnifiOS/EdgeOS/<insert name of Ubiquiti OS here> are all proprietary. So necessary details like "how do I write files into the operating system that persist" are all shrouded in secrecy and thus locked behind discovery tasks-- if they are even possible to begin with. From my brief spelunking, I concluded that the UXG Pro's OS is too dissimilar from the previous implementations that allowed NextDNS to be installed for that work to be re-usable.

All hope is not lost. If Ubiquiti looks charitably on NextDNS Inc.'s efforts, they might be willing to strike up a working relationship that would allow Ubiquiti to furnish NextDNS with enough details about the UXG Pro's OS to allow them to complete the port. But that certainly isn't something any of us outside contributors can instigate or participate in.

Or maybe someone with more discovery chops than I can do better at sussing out the UXG Pro's OS 🤷‍♂️

**Comment by joshuaspence** [comment]
There's some recent progress in https://github.com/boostchicken-dev/udm-utilities/issues/356 that might be relevant here.

**Comment by boostchicken** [comment]
Gonna steal my code again :).  

NOTE: I did not write the code to be clear for the UXG, thanks to the community!

**Comment by dotgrid** [comment]
@mbnn Michael, there has been some more progress here: https://github.com/joshuaspence/uxg-boot

I am using the above solution to run a custom DDNS container on my UXG Pro (just a quick test setup I tried).

(Side note: It looks possible to use _any_ of the `udm-utils` from the legendary @boostchicken. These previously couldn't run on the UXG, but John and the other contributors have figured that out now! 😃 😃 )

Regarding NextDNS, the community adding this as an option into `udm-utils` may be our best hope for running NextDNS on the UXG. 🤞

**Comment by boostchicken** [comment]
You could take my old nextdns docker file in udm-utilities and just update the nextdns version and deploy that image to the UXG like it says in the instructions.

**Comment by boostchicken** [comment]
I updated the docker image, instructions on how to do nextdns on uxg are here https://github.com/boostchicken-dev/udm-utilities/blob/main/nextdns/README.md

note: requires https://github.com/joshuaspence/uxg-boot

**Comment by mbnn** [comment]
Ok, ill try it out, but why would we want a isolated network stack for this? The nextdns cli simply uses the same stuff the UDM/UXG already has and then connects to it so it will use main default for DNS servince (Router dns).

Anyway so I tried to
First install the UXG Boot: https://github.com/unifi-utilities/uxg-boot worked,

- Then: # podman build . -t nextdns:latest
Error: error reading info about "/root/root/Dockerfile": stat /root/root/Dockerfile: no such file or directory

Any idea?

**Comment by mbnn** [comment]
> hey @mbnn, you are not in the folder that has the DOckerfile, make sure you are in the right directory, also, you can pull my prebuilt one. Logs of it's creation are available as an action in my repo so you can audit it's contents.

I'm sorry and I might be a noob here but I'm reading the documentation over here: https://github.com/unifi-utilities/unifios-utilities/blob/main/nextdns/README.md

It does not say anything about changing to a folder, so I dont understand what you mean with 
you are not in the folder that has the DOckerfile, make sure you are in the right directory, what directory exactly?

**Comment by joshuaspence** [comment]
@mbnn The easiest way would be to use the pre-built container like so:

```
podman create --detach --name nextdns --network host --restart always boostchicken/nextdns
podman start nextdns
```

**Comment by mbnn** [comment]
> @mbnn The easiest way would be to use the pre-built container like so:
> 
> ```
> podman create --detach --name nextdns --network host --restart always boostchicken/nextdns
> podman start nextdns
> ```

Yeah that works, and then?
# podman start nextdns
nextdns
# podman nextdns status

How to change options and find out if it works?

**Comment by joshuaspence** [comment]
I'm sorry, that's beyond my knowledge. I don't know anything about NextDNS. But now you have NextDNS up and running. How you configure it or use it I do not know exactly.

**Comment by joshuaspence** [comment]
I think you should just be able to follow https://github.com/nextdns/nextdns/wiki once you get a shell inside the nextdns container (`podman exec -it nextdns /bin/sh` or similar)

**Comment by KingJostle** [comment]
My friends, firmware UXG Pro 3.0.7 allowed me to install 
sh` -c 'sh -c "$(curl -sL https://nextdns.io/install)"'
without any workarounds, NextDNS after almost 2 years is now running on my UXG Pro!

**Comment by KingJostle** [comment]
> Hmmm, not so much with alpha fw 3.1.3 😀 '"clientName": "unknown"' when going to https://test.nextdns.io/
> 
> May 02 17:07:59 UXGPro nextdns[3031930]: Startup failed: proxy: tcp: listen tcp [::1]:53: bind: address already in use
> 
> Actually nvm, after a reboot of the UXG Pro it now works 😀
Thanks for the update, I’ll reboot mine after work too, as I have unidentified devices too showing in the log.

**Comment by mbnn** [comment]
@rs This seems to work now on UXG Pro after FW 3.0.x and 3.1.x.
The only strange thing is that client hostnames are not 100% correct, I think that might be related to https://github.com/nextdns/nextdns/issues/821

**Comment by rs** [comment]
Do you have an example of incorrectness?

**Comment by mbnn** [comment]
It seems to be working fine so far 😀 Thanks. on UXG Pro.

**Comment by dj-chalky** [comment]
> Hmmm, not so much with alpha fw 3.1.3 😀 '"clientName": "unknown"' when going to https://test.nextdns.io/
> 
> May 02 17:07:59 UXGPro nextdns[3031930]: Startup failed: proxy: tcp: listen tcp [::1]:53: bind: address already in use
> 
> Actually nvm, after a reboot of the UXG Pro it now works 😀

Are you doing anything specific when installing the app? I've tried setting the Router question to both Yes and No but each time I try and start NextDNS I get either the tcp or udp bind being already in use despite many reboots.

**Comment by mbnn** [comment]
No i just used the defaults, with Router I said yes. After that I rebooted and it worked.
However, since I dont 'trust' it keeps working I have removed it now and went for the normal DNS via IP in the UI.

**Comment by dj-chalky** [comment]
> No i just used the defaults, with Router I said yes. After that I rebooted and it worked. However, since I dont 'trust' it keeps working I have removed it now and went for the normal DNS via IP in the UI.

Thanks, I've ended up doing the same. I have access to a UDR and a UDM-Pro and both of those work fine, but they install differently and ask different questions leading to running on different Ports. Looks like the NextDNS Script doesn't identify a UXG properly and probably just sees it's as a Linux machine instead of a Router.

**Comment by jayyx2** [comment]
Hi all, I had an issue when installing on UXG-Pro.

`root@UXG-Pro:/run/dnsfilter# nextdns log
Jul 09 04:51:37 UXG-Pro systemd[1]: Started NextDNS DNS53 to DoH proxy..
Jul 09 04:51:37 UXG-Pro nextdns[1623]: Starting NextDNS 1.39.4/linux on :53
Jul 09 04:51:37 UXG-Pro nextdns[1623]: Starting mDNS discovery
Jul 09 04:51:37 UXG-Pro nextdns[1623]: Listening on TCP/:53
Jul 09 04:51:37 UXG-Pro nextdns[1623]: Listening on UDP/:53
Jul 09 04:51:37 UXG-Pro nextdns[1623]: Startup failed: proxy: tcp: listen tcp :53: bind: address already in use
Jul 09 04:51:37 UXG-Pro nextdns[1623]: Error: proxy: tcp: listen tcp :53: bind: address already in use
Jul 09 04:51:37 UXG-Pro systemd[1]: nextdns.service: Main process exited, code=exited, status=1/FAILURE
Jul 09 04:51:37 UXG-Pro systemd[1]: nextdns.service: Failed with result 'exit-code'.`

Reboots did not resolve this, the service could never start. I changed the settings in the DHCP DNS and this only partially works, too. Too force all queries to go through NextDNS, I had to modify some of the files on the system.

`vi /run/dnsfilter/dns-192.168.1.1-resolv.conf`
Here, I changed the name server to one of the IPs for NextDNS, and it magically began working. However, I encountered other problems the next day (unsure if related) I'm willing to help + experiment, but not sure what I can do to help resolve this open issue.
